### PR TITLE
Refactor TracerProvider to make configuration easy.

### PR DESCRIPTION
**Pull request by cijothomas**
Major refactoring in TracerProvider creation, incorporating feedback from https://github.com/open-telemetry/opentelemetry-dotnet/pull/1008, https://github.com/open-telemetry/opentelemetry-dotnet/pull/1008#discussion_r466734331, and offline feedbacks.

The examples are modified to show the new usage.

Note: This is fairly big change affecting 50 files so want to submit separate PRs for major feedbacks.

For significant contributions please make sure you have completed the following items:

* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: examples/Console/TestConsoleExporter.cs [hunk 1]
```diff
@@ -102,18 +101,6 @@ internal static object Run(ConsoleOptions options)
 
         internal class MyProcessor : ActivityProcessor
         {
```
Comment (cijothomas) [inline examples/Console/TestConsoleExporter.cs:103]: Writing own Processor is also greatly simplified now. No need of worrying about "next".

File: src/OpenTelemetry/Trace/TracerProviderBuilder.cs [hunk 2]
```diff
@@ -130,6 +158,108 @@ public TracerProviderBuilder AddActivitySource(string activitySourceName)
...
+            this.Sampler = this.Sampler ?? new ParentOrElseSampler(new AlwaysOnSampler());
+
+            var provider = new TracerProviderSdk
+            {
+                Resource = this.Resource,
+                Sampler = this.Sampler,
+                ActivityProcessor = this.ActivityProcessor,
+            };
+
+            var activitySource = new ActivitySourceAdapter(provider.Sampler, provider.ActivityProcessor, provider.Resource);
```
Comment (cijothomas) [inline src/OpenTelemetry/Trace/TracerProviderBuilder.cs:172]: activitysourceadapter won't fech ActiivtyProcessors added after building the provider. its a TODO on my list.

File: docs/trace/building-your-own-exporter/Program.cs [hunk 3]
```diff
@@ -24,12 +24,13 @@ public class Program
...
     {
-        using var tracerProvider = Sdk.CreateTracerProvider(
+        using var tracerProvider = Sdk.CreateTracerProviderBuilder()
+            .AddActivitySources(
             new string[]
             {
                 "MyCompany.MyProduct.MyLibrary",
             })
-            .AddMyExporter();
+            .Build().AddMyExporter();
```
Comment (reyang) [inline docs/trace/building-your-own-exporter/Program.cs:33]: I know you can add exporter after build, but hey 😺

File: docs/trace/building-your-own-exporter/Program.cs [hunk 4]
```diff
@@ -24,12 +24,13 @@ public class Program
 
     public static void Main()
     {
-        using var tracerProvider = Sdk.CreateTracerProvider(
+        using var tracerProvider = Sdk.CreateTracerProviderBuilder()
+            .AddActivitySources(
```
Comment (reyang) [inline docs/trace/building-your-own-exporter/Program.cs:28]: Trying to get more ideas - which one is more idiomatic? - `AddActivitySource` overload (singular form) or `AddActivitySources` (plural form).

Comment (eddynaka) [reply docs/trace/building-your-own-exporter/Program.cs:28]: Should people know that what we did so far is based on Activity? What if in the future it changes to something else? Should we remove the name Activity from it?

File: docs/trace/building-your-own-exporter/Program.cs (hunk 3 above)
Comment (ThomsonTan) [reply docs/trace/building-your-own-exporter/Program.cs:33]: Yes, for builder, it makes more sense to add everything before `Build()`.

File: docs/trace/building-your-own-exporter/Program.cs (hunk 4 above)
Comment (CodeBlanch) [reply docs/trace/building-your-own-exporter/Program.cs:28]: I feel like singular but with an overload for plural?
```csharp
void Add[Activity]Source(this TraceProviderBuilder builder, string sourceName);
void Add[Activity]Source(this TraceProviderBuilder builder, params string[] sourceNames);
```

File: docs/trace/building-your-own-sampler/Program.cs [hunk 5]
```diff
@@ -26,13 +26,15 @@ public class Program
...
+        using var tracerProvider = Sdk.CreateTracerProviderBuilder()
+            .AddActivitySources(
             new string[]
             {
                 "MyCompany.MyProduct.MyLibrary",
-            },
-            new MySampler())
-            .AddProcessor(new SimpleActivityProcessor(new ConsoleExporter(new ConsoleExporterOptions())));
+            })
+            .SetSampler(new MySampler())
```
Comment (CodeBlanch) [inline docs/trace/building-your-own-sampler/Program.cs:35]: "SetSampler" not "UseSampler"?

File: examples/Console/TestConsoleExporter.cs [hunk 6]
```diff
@@ -29,13 +29,12 @@ internal class TestConsoleExporter
...
-            using var tracerProvider = Sdk.CreateTracerProvider(
-                (builder) => builder.AddActivitySource("MyCompany.MyProduct.MyWebServer")
-                    .SetResource(Resources.CreateServiceResource("MyServiceName"))
-                    .UseConsoleExporter(
-                        opt => opt.DisplayAsJson = options.DisplayAsJson,
-                        (p) => p.AddProcessor((next) => new MyProcessor(next))));
+            // and use a custom MyProcessor, along with Console exporter.
+            using var tracerProvider = Sdk.CreateTracerProviderBuilder().AddActivitySource("MyCompany.MyProduct.MyWebServer")
+                .SetResource(Resources.CreateServiceResource("MyServiceName"))
+                .AddProcessor(new MyProcessor()) // This must be added before ConsoleExporter
```
Comment (alanwest) [inline examples/Console/TestConsoleExporter.cs:35]: What happens if it isn't added before ConsoleExporter?

Comment (cijothomas) [reply examples/Console/TestConsoleExporter.cs:35]: Any enrichments done by MyProcessor will not be visible to ConsoleExporter. Spec says processors are called in order they are added.

**Pull request merged**